feat: Angular 10 support

# Feature Request
Update Ionic framework to support Angular 10 and typescript 3.9.

**Ionic version:**
[x] **5.x**

**Describe the Feature Request**
The new version of Angular came out (Angular 10).
Update Ionic Angular dependencies to support the new version of Angular.
Angular 10 also pushes Typescript to version 3.9.

**Describe Preferred Solution**
Update Ionic Angular dependencies to support the new version of Angular.

**Additional Context**
Blog post about release: https://blog.angular.io/version-10-of-angular-now-available-78960babd41
Change Log v10: https://github.com/angular/angular/blob/master/CHANGELOG.md

> I did have an issue using the latest angular, ionic, and ionic-cli. I'm not certain if this is an issue that should be in the `ionic-cli` project or not as it appears to have to do with ionic's usage of angular schematics (let me know if this should become an issue in another repo):
> 
> ```shell
> $ ionic g page home
> > ng generate page home --project=app
> An unhandled exception occurred: Cannot find module '@schematics/angular/utility/project'
> Require stack:
> - /redacted/path/to/app/node_modules/@ionic/angular-toolkit/schematics/page/index.js
> - /redacted/path/to/app/node_modules/@angular-devkit/schematics/tools/export-ref.js
> - /redacted/path/to/app/node_modules/@angular-devkit/schematics/tools/index.js
> - /redacted/path/to/app/node_modules/@angular/cli/utilities/json-schema.js
> - /redacted/path/to/app/node_modules/@angular/cli/models/command-runner.js
> - /redacted/path/to/app/node_modules/@angular/cli/lib/cli/index.js
> - /redacted/path/to/app/node_modules/@angular/cli/lib/init.js
> - /redacted/path/to/app/node_modules/@angular/cli/bin/ng
> See "/tmp/ng-8qbJHz/angular-errors.log" for further details.
> [ERROR] Could not generate page.
> ```
> 
> And here's the output of `ionic info`:
> 
> ```shell
> Ionic:
> 
>    Ionic CLI                     : 6.10.1 (/redacted/path/to/node_modules/@ionic/cli)
>    Ionic Framework               : @ionic/angular 5.2.2
>    @angular-devkit/build-angular : 0.1000.0
>    @angular-devkit/schematics    : 10.0.0
>    @angular/cli                  : 10.0.0
>    @ionic/angular-toolkit        : 2.2.0
> 
> Cordova:
> 
>    Cordova CLI       : 9.0.0 (cordova-lib@9.0.1)
>    Cordova Platforms : not available
>    Cordova Plugins   : not available
> 
> Utility:
> 
>    cordova-res : 0.15.1
>    native-run  : 1.0.0
> 
> System:
> 
>    Android SDK Tools : 26.1.1 (/redacted/path/to/.android/sdk)
>    NodeJS            : v12.18.1 (/usr/bin/node)
>    npm               : 6.14.5
>    OS                : Linux 5.4
> ```

I am facing the same issue after upgrading from ionic 4 and angular 8.

Hello!

Catching error while running command `ng g component example`

# 🐞 Bug report

### Command (mark with an `x`)
<!-- Can you pin-point the command or commands that are effected by this bug? -->
<!-- ✍️edit: -->

- [ ] new
- [ ] build
- [ ] serve
- [ ] test
- [ ] e2e
- [x] generate
- [ ] add
- [ ] update
- [ ] lint
- [ ] xi18n
- [ ] run
- [ ] config
- [ ] help
- [ ] version
- [ ] doc


### Is this a regression?

<!-- Did this behavior use to work in the previous version? -->
<!-- ✍️--> Yes, the previous version in which this bug was not present was: Working well


## 🔬 Minimal Reproduction

<!-- ✍️--> 
After updating to `"@schematics/angular": "^10.0.3"`  command `ng g component example` breaks
in `"@schematics/angular": "^9.1.11"` everything is OK.


## 🔥 Exception or Error
<pre><code>An unhandled exception occurred: Cannot find module '@schematics/angular/utility/project'
Require stack:
- .../node_modules/@ionic/angular-toolkit/schematics/component/index.js
- .../node_modules/@angular-devkit/schematics/tools/export-ref.js
- .../node_modules/@angular-devkit/schematics/tools/index.js
- .../node_modules/@angular/cli/utilities/json-schema.js
- .../node_modules/@angular/cli/models/command-runner.js
- .../node_modules/@angular/cli/lib/cli/index.js
- .../.npm-global/lib/node_modules/@angular/cli/lib/init.js
- .../.npm-global/lib/node_modules/@angular/cli/bin/ng
See "/private/var/folders/2k/vlq4_8v57_98l_9rvxwmj1km0000gn/T/ng-aqdfcZ/angular-errors.log" for further details.
</code></pre>


## 🌍 Your Environment
<pre><code>Angular CLI: 10.0.3
Node: 12.16.3
OS: darwin x64

Angular: 10.0.4
... common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic
... router
Ivy Workspace: <error>

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1000.3 (cli-only)
@angular-devkit/build-angular   0.1000.3
@angular-devkit/core            10.0.3
@angular-devkit/schematics      10.0.3
@angular/cli                    10.0.3
@schematics/angular             10.0.3
@schematics/update              0.1000.3 (cli-only)
rxjs                            6.6.0
typescript                      3.9.7


ionic -v
6.10.1
</code></pre>

**Anything else relevant?**
<!-- ✍️Is this a browser specific issue? If so, please specify the browser and version. -->

<!-- ✍️Do any of these matter: operating system, IDE, package manager, HTTP server, ...? If so, please mention it below. -->

--- a/schematics/component/index.ts
+++ b/schematics/component/index.ts
@@ -1,9 +1,10 @@
 import { strings } from '@angular-devkit/core';
+import { ProjectDefinition } from '@angular-devkit/core/src/workspace';
 import { Rule, SchematicsException, Tree, apply, branchAndMerge, chain, filter, mergeWith, move, noop, template, url } from '@angular-devkit/schematics';
 import { buildRelativePath } from '@schematics/angular/utility/find-module';
 import { parseName } from '@schematics/angular/utility/parse-name';
-import { buildDefaultPath, getProject } from '@schematics/angular/utility/project';
 import { validateHtmlSelector, validateName } from '@schematics/angular/utility/validation';
+import { buildDefaultPath, getWorkspace } from '@schematics/angular/utility/workspace';
 import * as ts from 'typescript';
 
 import { buildSelector } from '../util';
@@ -123,21 +124,22 @@ function addImportToImports(host: Tree, options: ComponentOptions): void {
 }
 
 export default function(options: ComponentOptions): Rule {
-  return (host, context) => {
+  return async (host: Tree) => {
     if (!options.project) {
       throw new SchematicsException('Option (project) is required.');
     }
 
-    const project = getProject(host, options.project);
+    const workspace = await getWorkspace(host);
+    const project = workspace.projects.get(options.project);
 
     if (options.path === undefined) {
-      options.path = buildDefaultPath(project);
+      options.path = buildDefaultPath(project as ProjectDefinition);
     }
 
     const parsedPath = parseName(options.path, options.name);
     options.name = parsedPath.name;
     options.path = parsedPath.path;
-    options.selector = options.selector ? options.selector : buildSelector(options, project.prefix);
+    options.selector = options.selector ? options.selector : buildSelector(options, project?.prefix ?? 'app');
 
     validateName(options.name);
     validateHtmlSelector(options.selector);
@@ -158,6 +160,6 @@ export default function(options: ComponentOptions): Rule {
         addImportToNgModule(options),
         mergeWith(templateSource),
       ])),
-    ])(host, context);
+    ]);
   };
 }

--- a/schematics/page/index.ts
+++ b/schematics/page/index.ts
@@ -1,9 +1,10 @@
 import { Path, join, normalize, strings } from '@angular-devkit/core';
+import { ProjectDefinition } from '@angular-devkit/core/src/workspace';
 import { DirEntry, Rule, SchematicsException, Tree, apply, branchAndMerge, chain, filter, mergeWith, move, noop, template, url } from '@angular-devkit/schematics';
 import { ModuleOptions, buildRelativePath } from '@schematics/angular/utility/find-module';
 import { parseName } from '@schematics/angular/utility/parse-name';
-import { buildDefaultPath, getProject } from '@schematics/angular/utility/project';
 import { validateHtmlSelector, validateName } from '@schematics/angular/utility/validation';
+import { buildDefaultPath, getWorkspace } from '@schematics/angular/utility/workspace';
 import * as ts from 'typescript';
 
 import { buildSelector } from '../util';
@@ -12,19 +13,22 @@ import { Change, InsertChange } from '../util/change';
 
 import { Schema as PageOptions } from './schema';
 
-function findRoutingModuleFromOptions(host: Tree, options: ModuleOptions): Path | undefined {
+function findRoutingModuleFromOptions(
+  host: Tree,
+  options: ModuleOptions
+): Path | undefined {
   if (options.hasOwnProperty('skipImport') && options.skipImport) {
     return undefined;
   }
 
   if (!options.module) {
-    const pathToCheck = (options.path || '')
-                      + (options.flat ? '' : '/' + strings.dasherize(options.name));
+    const pathToCheck =
+      (options.path || '') +
+      (options.flat ? '' : '/' + strings.dasherize(options.name));
 
     return normalize(findRoutingModule(host, pathToCheck));
   } else {
-    const modulePath = normalize(
-      '/' + (options.path) + '/' + options.module);
+    const modulePath = normalize('/' + options.path + '/' + options.module);
     const moduleBaseName = normalize(modulePath).split('/').pop();
 
     if (host.exists(modulePath)) {
@@ -52,15 +56,19 @@ function findRoutingModule(host: Tree, generateDir: string): Path {
     if (matches.length === 1) {
       return join(dir.path, matches[0]);
     } else if (matches.length > 1) {
-      throw new Error('More than one module matches. Use skip-import option to skip importing '
-        + 'the component into the closest module.');
+      throw new Error(
+        'More than one module matches. Use skip-import option to skip importing ' +
+          'the component into the closest module.'
+      );
     }
 
     dir = dir.parent;
   }
 
-  throw new Error('Could not find an NgModule. Use the skip-import '
-    + 'option to skip importing in NgModule.');
+  throw new Error(
+    'Could not find an NgModule. Use the skip-import ' +
+      'option to skip importing in NgModule.'
+  );
 }
 
 function addRouteToNgModule(options: PageOptions): Rule {
@@ -78,19 +86,31 @@ function addRouteToNgModule(options: PageOptions): Rule {
     }
 
     const sourceText = text.toString('utf8');
-    const source = ts.createSourceFile(module, sourceText, ts.ScriptTarget.Latest, true);
+    const source = ts.createSourceFile(
+      module,
+      sourceText,
+      ts.ScriptTarget.Latest,
+      true
+    );
 
-    const pagePath = (
+    const pagePath =
       `/${options.path}/` +
       (options.flat ? '' : `${strings.dasherize(options.name)}/`) +
-      `${strings.dasherize(options.name)}.module`
-    );
+      `${strings.dasherize(options.name)}.module`;
 
     const relativePath = buildRelativePath(module, pagePath);
 
-    const routePath = strings.dasherize(options.routePath ? options.routePath : options.name);
+    const routePath = strings.dasherize(
+      options.routePath ? options.routePath : options.name
+    );
     const ngModuleName = `${strings.classify(options.name)}PageModule`;
-    const changes = addRouteToRoutesArray(source, module, routePath, relativePath, ngModuleName);
+    const changes = addRouteToRoutesArray(
+      source,
+      module,
+      routePath,
+      relativePath,
+      ngModuleName
+    );
     const recorder = host.beginUpdate(module);
 
     for (const change of changes) {
@@ -105,14 +125,24 @@ function addRouteToNgModule(options: PageOptions): Rule {
   };
 }
 
-function addRouteToRoutesArray(source: ts.SourceFile, ngModulePath: string, routePath: string, routeLoadChildren: string, ngModuleName: string): Change[] {
+function addRouteToRoutesArray(
+  source: ts.SourceFile,
+  ngModulePath: string,
+  routePath: string,
+  routeLoadChildren: string,
+  ngModuleName: string
+): Change[] {
   const keywords = findNodes(source, ts.SyntaxKind.VariableStatement);
 
   for (const keyword of keywords) {
     if (ts.isVariableStatement(keyword)) {
-      const [ declaration ] = keyword.declarationList.declarations;
+      const [declaration] = keyword.declarationList.declarations;
 
-      if (ts.isVariableDeclaration(declaration) && declaration.initializer && declaration.name.getText() === 'routes') {
+      if (
+        ts.isVariableDeclaration(declaration) &&
+        declaration.initializer &&
+        declaration.name.getText() === 'routes'
+      ) {
         const node = declaration.initializer.getChildAt(1);
         const lastRouteNode = node.getLastToken();
 
@@ -126,10 +156,20 @@ function addRouteToRoutesArray(source: ts.SourceFile, ngModulePath: string, rout
         if (lastRouteNode.kind === ts.SyntaxKind.CommaToken) {
           trailingCommaFound = true;
         } else {
-          changes.push(new InsertChange(ngModulePath, lastRouteNode.getEnd(), ','));
+          changes.push(
+            new InsertChange(ngModulePath, lastRouteNode.getEnd(), ',')
+          );
         }
 
-        changes.push(new InsertChange(ngModulePath, lastRouteNode.getEnd() + 1, `  {\n    path: '${routePath}',\n    loadChildren: () => import('${routeLoadChildren}').then( m => m.${ngModuleName})\n  }${trailingCommaFound ? ',' : ''}\n`));
+        changes.push(
+          new InsertChange(
+            ngModulePath,
+            lastRouteNode.getEnd() + 1,
+            `  {\n    path: '${routePath}',\n    loadChildren: () => import('${routeLoadChildren}').then( m => m.${ngModuleName})\n  }${
+              trailingCommaFound ? ',' : ''
+            }\n`
+          )
+        );
 
         return changes;
       }
@@ -140,15 +180,16 @@ function addRouteToRoutesArray(source: ts.SourceFile, ngModulePath: string, rout
 }
 
 export default function(options: PageOptions): Rule {
-  return (host, context) => {
+  return async (host: Tree) => {
     if (!options.project) {
       throw new SchematicsException('Option (project) is required.');
     }
 
-    const project = getProject(host, options.project);
+    const workspace = await getWorkspace(host);
+    const project = workspace.projects.get(options.project);
 
     if (options.path === undefined) {
-      options.path = buildDefaultPath(project);
+      options.path = buildDefaultPath(project as ProjectDefinition);
     }
 
     options.module = findRoutingModuleFromOptions(host, options);
@@ -156,7 +197,7 @@ export default function(options: PageOptions): Rule {
     const parsedPath = parseName(options.path, options.name);
     options.name = parsedPath.name;
     options.path = parsedPath.path;
-    options.selector = options.selector ? options.selector : buildSelector(options, project.prefix);
+    options.selector = options.selector ? options.selector : buildSelector(options, project?.prefix ?? 'app');
 
     validateName(options.name);
     validateHtmlSelector(options.selector);
@@ -165,17 +206,16 @@ export default function(options: PageOptions): Rule {
       options.spec ? noop() : filter(p => !p.endsWith('.spec.ts')),
       template({
         ...strings,
-        'if-flat': (s: string) => options.flat ? '' : s,
+        'if-flat': (s: string) => (options.flat ? '' : s),
         ...options,
       }),
       move(parsedPath.path),
     ]);
 
     return chain([
-      branchAndMerge(chain([
-        addRouteToNgModule(options),
-        mergeWith(templateSource),
-      ])),
-    ])(host, context);
+      branchAndMerge(
+        chain([addRouteToNgModule(options), mergeWith(templateSource)])
+      ),
+    ]);
   };
 }

